docs/architecture/domain.md

```diff
@@ -0,0 +1,117 @@
+
+### Step
+
+A step is a named action that needs to be performed, usually a single shell command. If the command finishes with a non-zero exit code, the step is considered a failure.
+
+### Container
+
+A Container runs a [task] in an isolated environment. This is done in order to determine dependency version compatibility, without affecting other tasks that may be running at the some time. This is implemented using Docker containers.
+
+### Job
+
+A job consists of executing multiple sequential [steps] inside a specified [container]. If any step in the series fails, then the entire job is considered failed and subsequent steps will be skipped (unless configured otherwise).
+
+Jobs work by checking out the [source code] to a specified commit, setting the desired environment variables, and executing the specified [steps].
+
+During the job, the executing [steps] share three pieces of context:
+
+ - Filesystem
+ - [Container]
+ - [Metadata]
+
+Jobs are started by changes in [source code], as a trigger from the [workflow], or by human decision.
+
+### Matrix
+
+A build matrix is a set of environment permutations for a given [job]. Permutations are usually used for testing against multiple [container]s or test types.
+
+In this example, the job will run 4 times for a given [build]:
+```yaml
+matrix:
+    NODE_VERSION:
+        - 4
+        - 6
+    EXPRESS_VERSION:
+        - 2
+        - 3
+```
+
+ - `NODE_VERSION=4` and `EXPRESS_VERSION=2`
+ - `NODE_VERSION=4` and `EXPRESS_VERSION=3`
+ - `NODE_VERSION=6` and `EXPRESS_VERSION=2`
+ - `NODE_VERSION=6` and `EXPRESS_VERSION=3`
+
+### Workflow
+
+Workflow is the defined order that [job]s will execute in after a successful [build] of the `main` job on the default branch. Jobs can be executed in parallel, series, or a combination of the two to allow for all possibilities. Workflow must contain all defined jobs in the pipeline.  All jobs executed in a given workflow run share:
+
+ - Source code checked out from the same git commit from the `main` [build] that triggered this
+ - A copy of the [metadata] from the `main` [build] that triggered this
+
+In the following example of a workflow section, this is the flow:
+```yaml
+workflow:
+    - publish
+    - parallel:
+        - series:
+            - deploy-west
+            - validate-west
+        - series:
+            - deploy-east
+            - validate-east
+```
+
+After the merge of a pull-request to master:
+
+ - `main` will run and trigger `publish`
+ - `publish` will trigger `deploy-west` and `deploy-east` in parallel
+ - `deploy-west` triggers `validate-west`
+ - `deploy-east` triggers `validate-east`
+
+### Pipeline
+
+A pipeline represents a collection of [job]s that share the same source code. These jobs are executed in the order defined by the [workflow]. The `main` job is required to be defined in every pipeline:
```

Comment:
`[job]s` looks kinda weird. Can we add `[jobs]` and just add that to point to the same link?